tfrecord/reader.py

```diff
@@ -87,24 +91,40 @@ def tfrecord_loader(data_path, index_path, description, shard=None):
         example = example_pb2.Example()
         example.ParseFromString(record)
 
+        all_keys = list(example.features.feature.keys())
+        if description is None:
+            description = dict.fromkeys(all_keys, None)
+        elif isinstance(description, list):
+            description = dict.fromkeys(description, None)
+
         features = {}
         for key, typename in description.items():
-            tf_typename = {
-                "byte": "bytes_list",
-                "float": "float_list",
-                "int": "int64_list"
-            }[typename]
-            if key not in example.features.feature:
-                raise ValueError("Key {} doesn't exist.".format(key))
-            value = getattr(example.features.feature[key], tf_typename).value
-            if typename == "byte":
+            if key not in all_keys:
+                raise KeyError(f"Key {key} doesn't exist (select from {all_keys})!")
+            # NOTE: We assume that each key in the example has only one field
+            # (either "bytes_list", "float_list", or "int64_list")!
+            field = example.features.feature[key].ListFields()[0]
+            inferred_typename, value = field[0].name, field[1].value
+            if typename is not None:
+                typename_mapping = {
+                    "byte": "bytes_list",
+                    "float": "float_list",
+                    "int": "int64_list"
+                }
+                tf_typename = typename_mapping[typename]
+                if tf_typename != inferred_typename:
+                    reversed_mapping = {v:k for k, v in typename_mapping.items()}
+                    raise TypeError(f"Incompatible type '{typename}' for {key} "
+                                    f"(should be {reversed_mapping[inferred_typename]}).")
```

Comment:
This implementation feels a bit "dirty" to me, any suggestions for improvement? For example, would it be ok to instead assert that the typenames don't match, without raising a `TypeError`,

```python
tf_typename = {
    "byte": "bytes_list",
    "float": "float_list",
    "int": "int64_list"
}[typename]
assert inferred_typename == tf_typename
```

Reply:
In python asserts are pretty useless. It should only be used for debugging not in production.